overlord/devicestate/handlers_install.go

```diff
@@ -1363,9 +1416,51 @@ func (m *DeviceManager) doInstallSetupStorageEncryption(t *state.Task, _ *tomb.T
 	if err := t.Get("on-volumes", &onVolumes); err != nil {
 		return err
 	}
-	logger.Debugf("install-setup-storage-encyption for %q on %v", systemLabel, onVolumes)
-	// TODO: find device with role system-{data,seed} and setup
-	// storage encryption
+	logger.Debugf("install-setup-storage-encryption for %q on %v", systemLabel, onVolumes)
+
+	st.Unlock()
+	sys, snapInfos, snapSeeds, mntPtForType, unmount, err := m.loadAndMountSystemLabelSnaps(systemLabel)
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	defer unmount()
+
+	// Gadget information
+	snapf, err := snapfile.Open(snapSeeds[snap.TypeGadget].Path)
+	if err != nil {
+		return fmt.Errorf("cannot open gadget snap: %v", err)
+	}
+	gadgetInfo, err := gadget.ReadInfoFromSnapFileNoValidate(snapf, sys.Model)
+	if err != nil {
+		return fmt.Errorf("reading gadget information: %v", err)
+	}
+
+	encryptInfo, err := m.encryptionSupportInfo(sys.Model, snapInfos[snap.TypeKernel], gadgetInfo)
+	if err != nil {
+		return err
+	}
+	if encryptInfo.Type == secboot.EncryptionTypeNone {
+		return fmt.Errorf("encryption unavailable on this device")
+	}
 
-	return fmt.Errorf("setup storage encryption step not implemented yet")
+	// FIXME this is fixed to LUKS at the moment
+	encryptionSetupData, err := installEncryptPartitions(onVolumes, mntPtForType[snap.TypeGadget],
+		mntPtForType[snap.TypeKernel], sys.Model, secboot.EncryptionTypeLUKS, perfTimings)
+	if err != nil {
+		return err
+	}
+
+	// Store created devices in the change so they can be accessed from the installer
+	apiData := make(map[string]interface{})
```

Comment:
we probably want a level in between instead of just the roles directly. Something like "encrypted-devices"